ah ok but its not the race im talking about:

```
---> Making bundle: test-integration (in bundles/1.4.1-dev/test-integration)
Created binary: /var/cache/drone/src/github.com/docker/docker/bundles/1.4.1-dev/test-integration/dockerinit-1.4.1-dev
+ go test -test.timeout=30m github.com/docker/docker/integration
time="2015-01-31T19:29:11Z" level="info" msg="enableIPv6 = false" 
time="2015-01-31T19:29:11Z" level="info" msg="WARNING: Your kernel does not support cgroup swap limit." 
Status: Image is up to date for docker-test-image
time="2015-01-31T19:29:12Z" level="info" msg="enableIPv6 = false" 
time="2015-01-31T19:29:12Z" level="info" msg="WARNING: Your kernel does not support cgroup swap limit." 
time="2015-01-31T19:29:13Z" level="info" msg="Listening for HTTP on tcp (127.0.0.1:4270)" 
time="2015-01-31T19:29:14Z" level="info" msg="enableIPv6 = false" 
time="2015-01-31T19:29:14Z" level="info" msg="WARNING: Your kernel does not support cgroup swap limit." 
time="2015-01-31T19:29:14Z" level="info" msg="Listening for HTTP on tcp (localhost:4271)" 
time="2015-01-31T19:29:15Z" level="info" msg="enableIPv6 = false" 
time="2015-01-31T19:29:15Z" level="info" msg="WARNING: Your kernel does not support cgroup swap limit." 
time="2015-01-31T19:29:16Z" level="info" msg="Listening for HTTP on tcp (localhost:4272)" 
time="2015-01-31T19:29:17Z" level="info" msg="enableIPv6 = false" 
time="2015-01-31T19:29:17Z" level="info" msg="WARNING: Your kernel does not support cgroup swap limit." 
time="2015-01-31T19:29:17Z" level="error" msg="Handler for GET /images/{name:.*}/get returned error: could not find image: no such id: 83599e29c455eb719f77d799bc7c51521b9551972f5a850d7ad265bc1b5292f6" 
time="2015-01-31T19:29:17Z" level="error" msg="HTTP Error: statusCode=404 could not find image: no such id: 83599e29c455eb719f77d799bc7c51521b9551972f5a850d7ad265bc1b5292f6" 
--- FAIL: TestSaveImageAndThenLoad (0.44s)
	api_test.go:36: 200 OK expected, received 404
time="2015-01-31T19:29:17Z" level="info" msg="enableIPv6 = false" 
time="2015-01-31T19:29:17Z" level="info" msg="WARNING: Your kernel does not support cgroup swap limit." 
--- FAIL: TestGetContainersTop (0.43s)
	utils_test.go:51: No such image: 83599e29c455eb719f77d799bc7c51521b9551972f5a850d7ad265bc1b5292f6 (tag: latest)
time="2015-01-31T19:29:17Z" level="info" msg="enableIPv6 = false" 
time="2015-01-31T19:29:18Z" level="info" msg="WARNING: Your kernel does not support cgroup swap limit." 
--- FAIL: TestPostCommit (0.41s)
	utils_test.go:51: No such image: 83599e29c455eb719f77d799bc7c51521b9551972f5a850d7ad265bc1b5292f6 (tag: latest)
time="2015-01-31T19:29:18Z" level="info" msg="enableIPv6 = false" 
time="2015-01-31T19:29:18Z" level="info" msg="WARNING: Your kernel does not support cgroup swap limit." 
time="2015-01-31T19:29:18Z" level="error" msg="Handler for POST /containers/create returned error: No such image: 83599e29c455eb719f77d799bc7c51521b9551972f5a850d7ad265bc1b5292f6 (tag: latest)" 
time="2015-01-31T19:29:18Z" level="error" msg="HTTP Error: statusCode=404 No such image: 83599e29c455eb719f77d799bc7c51521b9551972f5a850d7ad265bc1b5292f6 (tag: latest)" 
--- FAIL: TestPostContainersCreate (0.44s)
	utils_test.go:130: Unexpected http error: 404
time="2015-01-31T19:29:18Z" level="info" msg="enableIPv6 = false" 
time="2015-01-31T19:29:18Z" level="info" msg="WARNING: Your kernel does not support cgroup swap limit." 
time="2015-01-31T19:29:19Z" level="error" msg="Error parsing media type:  error: mime: no media type" 
time="2015-01-31T19:29:19Z" level="error" msg="Handler for POST /containers/create returned error: Content-Type specified () must be 'application/json'" 
time="2015-01-31T19:29:19Z" level="error" msg="HTTP Error: statusCode=500 Content-Type specified () must be 'application/json'" 
time="2015-01-31T19:29:19Z" level="info" msg="enableIPv6 = false" 
time="2015-01-31T19:29:19Z" level="info" msg="WARNING: Your kernel does not support cgroup swap limit." 
time="2015-01-31T19:29:19Z" level="error" msg="Handler for POST /containers/create returned error: No such image: 83599e29c455eb719f77d799bc7c51521b9551972f5a850d7ad265bc1b5292f6 (tag: latest)" 
time="2015-01-31T19:29:19Z" level="error" msg="HTTP Error: statusCode=404 No such image: 83599e29c455eb719f77d799bc7c51521b9551972f5a850d7ad265bc1b5292f6 (tag: latest)" 
--- FAIL: TestPostCreateNull (0.29s)
	utils_test.go:130: Unexpected http error: 404
time="2015-01-31T19:29:19Z" level="info" msg="enableIPv6 = false" 
time="2015-01-31T19:29:19Z" level="info" msg="WARNING: Your kernel does not support cgroup swap limit." 
--- FAIL: TestPostContainersKill (0.18s)
	utils_test.go:51: No such image: 83599e29c455eb719f77d799bc7c51521b9551972f5a850d7ad265bc1b5292f6 (tag: latest)
time="2015-01-31T19:29:19Z" level="info" msg="enableIPv6 = false" 
time="2015-01-31T19:29:19Z" level="info" msg="WARNING: Your kernel does not support cgroup swap limit." 
--- FAIL: TestPostContainersRestart (0.19s)
	utils_test.go:51: No such image: 83599e29c455eb719f77d799bc7c51521b9551972f5a850d7ad265bc1b5292f6 (tag: latest)
time="2015-01-31T19:29:19Z" level="info" msg="enableIPv6 = false" 
time="2015-01-31T19:29:20Z" level="info" msg="WARNING: Your kernel does not support cgroup swap limit." 
--- FAIL: TestPostContainersStart (0.39s)
	utils_test.go:51: No such image: 83599e29c455eb719f77d799bc7c51521b9551972f5a850d7ad265bc1b5292f6 (tag: latest)
time="2015-01-31T19:29:20Z" level="info" msg="enableIPv6 = false" 
time="2015-01-31T19:29:20Z" level="info" msg="WARNING: Your kernel does not support cgroup swap limit." 
--- FAIL: TestPostContainersStop (0.41s)
	utils_test.go:51: No such image: 83599e29c455eb719f77d799bc7c51521b9551972f5a850d7ad265bc1b5292f6 (tag: latest)
time="2015-01-31T19:29:20Z" level="info" msg="enableIPv6 = false" 
time="2015-01-31T19:29:20Z" level="info" msg="WARNING: Your kernel does not support cgroup swap limit." 
--- FAIL: TestPostContainersWait (0.40s)
	utils_test.go:51: No such image: 83599e29c455eb719f77d799bc7c51521b9551972f5a850d7ad265bc1b5292f6 (tag: latest)
time="2015-01-31T19:29:21Z" level="info" msg="enableIPv6 = false" 
time="2015-01-31T19:29:21Z" level="info" msg="WARNING: Your kernel does not support cgroup swap limit." 
--- FAIL: TestPostContainersAttach (0.39s)
	utils_test.go:51: No such image: 83599e29c455eb719f77d799bc7c51521b9551972f5a850d7ad265bc1b5292f6 (tag: latest)
time="2015-01-31T19:29:21Z" level="info" msg="enableIPv6 = false" 
time="2015-01-31T19:29:21Z" level="info" msg="WARNING: Your kernel does not support cgroup swap limit." 
--- FAIL: TestPostContainersAttachStderr (0.41s)
	utils_test.go:51: No such image: 83599e29c455eb719f77d799bc7c51521b9551972f5a850d7ad265bc1b5292f6 (tag: latest)
time="2015-01-31T19:29:21Z" level="info" msg="enableIPv6 = false" 
time="2015-01-31T19:29:22Z" level="info" msg="WARNING: Your kernel does not support cgroup swap limit." 
time="2015-01-31T19:29:22Z" level="info" msg="enableIPv6 = false" 
time="2015-01-31T19:29:22Z" level="info" msg="WARNING: Your kernel does not support cgroup swap limit." 
time="2015-01-31T19:29:22Z" level="info" msg="enableIPv6 = false" 
time="2015-01-31T19:29:22Z" level="info" msg="WARNING: Your kernel does not support cgroup swap limit." 
--- FAIL: TestDeleteImages (0.39s)
	api_test.go:761: could not find image: no such id: docker-test-image
time="2015-01-31T19:29:23Z" level="info" msg="enableIPv6 = false" 
time="2015-01-31T19:29:23Z" level="info" msg="WARNING: Your kernel does not support cgroup swap limit." 
--- FAIL: TestPostContainersCopy (0.43s)
	utils_test.go:51: No such image: 83599e29c455eb719f77d799bc7c51521b9551972f5a850d7ad265bc1b5292f6 (tag: latest)
time="2015-01-31T19:29:23Z" level="info" msg="enableIPv6 = false" 
time="2015-01-31T19:29:23Z" level="info" msg="WARNING: Your kernel does not support cgroup swap limit." 
time="2015-01-31T19:29:23Z" level="error" msg="No such container: id_not_found" 
time="2015-01-31T19:29:23Z" level="info" msg="enableIPv6 = false" 
time="2015-01-31T19:29:24Z" level="info" msg="WARNING: Your kernel does not support cgroup swap limit." 
time="2015-01-31T19:29:24Z" level="error" msg="Handler for POST /containers/create returned error: No such image: docker-test-image (tag: latest)" 
time="2015-01-31T19:29:24Z" level="error" msg="HTTP Error: statusCode=404 No such image: docker-test-image (tag: latest)" 
--- FAIL: TestConstainersStartChunkedEncodingHostConfig (0.44s)
	utils_test.go:130: Unexpected http error: 404
time="2015-01-31T19:29:24Z" level="info" msg="POST /v1.17/containers/create" 
time="2015-01-31T19:29:24Z" level="error" msg="Handler for POST /containers/create returned error: No such image: 83599e29c455eb719f77d799bc7c51521b9551972f5a850d7ad265bc1b5292f6 (tag: latest)" 
time="2015-01-31T19:29:24Z" level="error" msg="HTTP Error: statusCode=404 No such image: 83599e29c455eb719f77d799bc7c51521b9551972f5a850d7ad265bc1b5292f6 (tag: latest)" 
--- FAIL: TestRunDetach (10.00s)
	commands_test.go:89: Waiting for the container to be started timed out
time="2015-01-31T19:29:34Z" level="info" msg="POST /v1.17/containers/create" 
time="2015-01-31T19:29:34Z" level="error" msg="Handler for POST /containers/create returned error: No such image: 83599e29c455eb719f77d799bc7c51521b9551972f5a850d7ad265bc1b5292f6 (tag: latest)" 
time="2015-01-31T19:29:34Z" level="error" msg="HTTP Error: statusCode=404 No such image: 83599e29c455eb719f77d799bc7c51521b9551972f5a850d7ad265bc1b5292f6 (tag: latest)" 
--- FAIL: TestAttachDetach (10.00s)
	commands_test.go:181: Invalid repository name (83599e29c455eb719f77d799bc7c51521b9551972f5a850d7ad265bc1b5292f6), cannot specify 64-byte hexadecimal strings
	commands_test.go:89: Waiting for the container to be started timed out
time="2015-01-31T19:29:44Z" level="info" msg="POST /v1.17/containers/create" 
time="2015-01-31T19:29:44Z" level="error" msg="Handler for POST /containers/create returned error: No such image: 83599e29c455eb719f77d799bc7c51521b9551972f5a850d7ad265bc1b5292f6 (tag: latest)" 
time="2015-01-31T19:29:44Z" level="error" msg="HTTP Error: statusCode=404 No such image: 83599e29c455eb719f77d799bc7c51521b9551972f5a850d7ad265bc1b5292f6 (tag: latest)" 
--- FAIL: TestAttachDetachTruncatedID (2.00s)
	commands_test.go:269: Invalid repository name (83599e29c455eb719f77d799bc7c51521b9551972f5a850d7ad265bc1b5292f6), cannot specify 64-byte hexadecimal strings
	commands_test.go:89: Starting container timed out
time="2015-01-31T19:29:46Z" level="info" msg="POST /v1.17/containers/create" 
time="2015-01-31T19:29:46Z" level="error" msg="Handler for POST /containers/create returned error: No such image: 83599e29c455eb719f77d799bc7c51521b9551972f5a850d7ad265bc1b5292f6 (tag: latest)" 
time="2015-01-31T19:29:46Z" level="error" msg="HTTP Error: statusCode=404 No such image: 83599e29c455eb719f77d799bc7c51521b9551972f5a850d7ad265bc1b5292f6 (tag: latest)" 
--- FAIL: TestAttachDisconnect (10.00s)
	commands_test.go:89: Waiting for CmdRun timed out
time="2015-01-31T19:29:56Z" level="info" msg="enableIPv6 = false" 
time="2015-01-31T19:29:56Z" level="info" msg="WARNING: Your kernel does not support cgroup swap limit." 
time="2015-01-31T19:29:56Z" level="fatal" msg="Test image 83599e29c455eb719f77d799bc7c51521b9551972f5a850d7ad265bc1b5292f6 not found in /var/lib/docker/tmp/docker-testd728-newTestEngine-632342385/graph: map[]" 
exit status 1
FAIL	_/var/cache/drone/src/github.com/docker/docker/integration	45.251s
```

--- a/graph/pull.go
+++ b/graph/pull.go
@@ -387,6 +387,9 @@ func (s *TagStore) pullV2Repository(eng *engine.Engine, r *registry.Session, out
 		if err != nil {
 			return err
 		}
+		if len(tags) == 0 {
+			return registry.ErrDoesNotExist
+		}
 		for _, t := range tags {
 			if downloaded, err := s.pullV2Tag(eng, r, out, endpoint, repoInfo, t, sf, parallel, auth); err != nil {
 				return err

--- a/registry/session_v2.go
+++ b/registry/session_v2.go
@@ -128,6 +128,8 @@ func (r *Session) HeadV2ImageBlob(ep *Endpoint, imageName, sumType, sum string, 
 	case res.StatusCode >= 200 && res.StatusCode < 400:
 		// return something indicating no push needed
 		return true, nil
+	case res.StatusCode == 401:
+		return false, errLoginRequired
 	case res.StatusCode == 404:
 		// return something indicating blob push needed
 		return false, nil

--- a/registry/token.go
+++ b/registry/token.go
@@ -51,10 +51,12 @@ func getToken(username, password string, params map[string]string, registryEndpo
 		reqParams.Add("scope", scopeField)
 	}
 
-	reqParams.Add("account", username)
+	if username != "" {
+		reqParams.Add("account", username)
+		req.SetBasicAuth(username, password)
+	}
 
 	req.URL.RawQuery = reqParams.Encode()
-	req.SetBasicAuth(username, password)
 
 	resp, err := client.Do(req)
 	if err != nil {

